sorting/count_inversions.cpp

```diff
@@ -0,0 +1,194 @@
+}
+
+/**
+ * 
+ * The mergeSort() function implements Merge Sort, a
+ * Divide and conquer algorithm, it divides the input
+ * array into two halves and calls itself for each 
+ * sub-array and then calls the merge() function to 
+ * merge the two halves.
+ * 
+ * @param arr   - array to be sorted
+ * @param temp  - merged resultant array
+ * @param left  - lower bound of array
+ * @param right - upper bound of array
+ * 
+ */
+int mergeSort(int* arr, int* temp, int left, int right) {
+    int mid = 0, inv_count = 0;
+    if(right>left)
+    {
+        // midpoint to split the array
+        mid = (right + left) / 2;
+        // Add inversions in left and right sub-arrays
+        inv_count += mergeSort(arr, temp, left, mid);   // left sub-array
+        inv_count += mergeSort(arr, temp, mid+1, right);
+
+        // inversions in the merge step
+        inv_count += merge(arr, temp, left, mid, right);
+    }
+    return inv_count;
+}
+
+/**
+ * Function to merge two sub-arrays. merge() function is called 
+ * from mergeSort() to merge the array after it split for sorting 
+ * by the mergeSort() funtion.
+ * 
+ * In this case the merge fuction will also count and return 
+ * inversions detected when merging the sub arrays.
+ * 
+ * @param arr    input array, data-menber of vector
+ * @param temp   stores the resultant merged array
+ * @param left   lower bound of arr[] and left-sub-array
+ * @param mid    midpoint, upper bound of left sub-array,
+ *               (mid+1) gives the lower bound of right-sub-array
+ * @param right  upper bound of arr[] and right-sub-array
+ * 
+ */
+int merge(int* arr, int* temp, int left, int mid, int right)    {
+    int i = left;   /* i --> index of left sub-array */
+    int j = mid+1;  /* j --> index for right sub-array */
+    int k = left;   /* k --> index for resultant array temp */
+    int inv_count = 0;  // inversion count
+
+    while((i <= mid) && (j <= right))
+    {
+        if (arr[i] <= arr[j]) {
+            temp[k++] = arr[i++];
+        }
+        else{
+            temp[k++] = arr[j++];
+            inv_count += (mid - i + 1); // tricky; may vary depending on selection of sub-array
+        }
+    } 
+    // Add remaining elements from the larger subarray to the end of temp
+    while( i <= mid) {
+        temp[k++] = arr[i++];
+    }
+    while( j <= right) {
+        temp[k++] = arr[j++];
+    }
+    // Copy temp[] to arr[]
+    for( k=left; k<=right; k++) {
+        arr[k] = temp[k];
+    }
+    return inv_count;
+}
+
+/**
+ * UTILITY function to print array.
+ * @param arr[]   array to print
+ * @param array_size    size of input array arr[]
+ * 
+ */
+void show(int* arr, const int array_size) {
+    std::cout<< "Printing array: \n";
+    for( int i=0; i<array_size; i++){
+        std::cout<<" "<< arr[i];
+    }
+    std::cout << "\n";
+}
+
+}   // namespace inversion
+}   // namespace sorting
+
+using sorting::inversion::countInversion;
+using sorting::inversion::show;
```

Comment:
Please remove this and use namespace directly.

Reply:
Using namespace directly generates clang-tidy errors

Reply:
If you're going to call a function inside those namespaces, you need to add `sorting::inversion` before the function's name.

Reply:
So i need to remove the `using` directives and declare and call functions as <namespace>::<function_name>( args ) ?

Reply:
When registering a function inside a namespace, you don't need to add the `sorting::inversion` at the beginning.
When calling the function itself, it **is** needed to add that at the beginning. Otherwise, it won't work.

Example:

```c++
// Registering namespace
namespace sorting {
void function(int arg_1, int arg_2) {/** definition */ }
} // namespace sorting

int main() {
    int arg_1, arg_2 = 0;
    sorting::function(arg_1, arg_2); // Calling the function inside a namespace
}
```

Reply:
Noted. Fixing all in the next commit.